KEYCLOAK-14209 Client policies admin console support. Changing of format of JSON for client policies and profiles. Remove support for default policies

mposolda opened the pull request:
This PR is still draft and it has some things, which should be addressed before it is merged. However I am sending PR to receive some feedback for the UI and functionality.

Changes in this PR:

- Added support in the old admin console for client policies and client profiles
-- I've tried to add admin console based on the proposed designs for the new admin console - TODO . However it is not exactly as proposed in the design due some limitations of the old admin console.
-- Everything is currently available in the realm "Client Policies" tab under realm tabs
-- Both "Form view" and "JSON view" available as proposed in the design for both client profiles and client policies

- Added some changes in the JSON format and added some more type-safety to the source-code to avoid some manual JSON parsing whenever possible. However there is a chance that format still may need to be changed based on the feedback from the mailing list thread https://groups.google.com/g/keycloak-dev/c/MfoTQbNPljE .

- Some minor refactoring in providers to avoid some boilerplate code. For example added AbstractClientPolicyConditionProvider and moved method "isNegativeLogic" to this class to avoid repeating it in all the condition provider implementation classes.

- ClientUpdateSourceRolesCondition was changed a bit to support reference roles in same format like our other providers in various SPIs (For example HardcodedRoleMapper). That means realm roles with name "foo" is in the configuration referenced as "foo" and the client role "bar" of client with client ID "my-client" is referenced as "my-client.bar".

There are still some things, which needs to be fixed before this PR is merged (For example align configuration format based on the feedback from mailing list, make sure that "Client Policies" tab is available just when Client policy feature enabled etc). But I am sending PR earlier to collect feedback.

mposolda committed:
KEYCLOAK-14209 Client policies admin console support. Small changing of format of JSON for client policies and profiles. Refactoring

mposolda commented:
@tnorimat Do you please have a chance to take a look at this PR?

tnorimat commented:
@mposolda Yes, I will look at this PR and add comments.

mposolda committed:
KEYCLOAK-14209 KEYCLOAK-17988 Refactoring admin console and code based on feedback. Removing builtin policies

vmuzikar reviewed (commented):
Just a few notes regarding the UI (haven't yet checked the code line by line).

* The executors UI is not very user friendly (I was quite confused myself at first :)). We need to display some more descriptive names of the executors instead of provider ID. Something similar to mappers for example.
* Do we need the JSON view? IMHO it doesn't fit with the rest of the admin console. What is the use case for this? Do we expect the users to enter JSON directly? I'd say an import/export button would be a better alternative (like we do it for Clients e.g.). And if decide to input JSON through UI, we should do it properly and at least use some lib for JSON syntax highlighting and ideally with autocomplete as well.
* If we keep the JSON view, we should probably improve the syntax error handling. Telling the user to open dev tools is not very user friendly. :) Maybe a modal dialog with more explicit error message would be better.
* Do we need built-in profiles? I suppose we do (FAPI and stuff), right? It looks a bit weird, the empty actions in the UI. :) Is it possible to enable/disable the built-in profiles? If yes, we could perhaps have a button for it in the actions column.
* The executors for built-in profiles are read-only. This is of course correct. However, once you click it, there's no "back" button or something. It's a dead end. :)
* Maybe we should have some bread-crumb navigation, like we have for clients and mappers e.g. It seems especially confusing for executors and conditions, when all the navigation (the Realm settings tabs) disappear. So maybe bread-crumb in this case could be somewhat helpful.
* Just a nitpick. The "Realm Settings" button in the left nav bar is not highlighted as active.
* It'd be a nice (and easy) improvement if client profiles could be opened directly from client policies. The linked profiles are currently not clickable.
* I'm afraid we need some UI tests.

I know this is just a draft and you may (or may not :)) already have plans to fix/improve things I mentioned. And I'm also not sure if we even have time for stuff like that, or how good we want the UI to be. :) So please take it just as a recommendation/personal opinion from me. :)

mposolda commented:
> Just a few notes regarding the UI (haven't yet checked the code line by line).
> 
>     * The executors UI is not very user friendly (I was quite confused myself at first :)). We need to display some more descriptive names of the executors instead of provider ID. Something similar to mappers for example.
> 

+1, Will it be possible to create follow-up JIRA for it under https://issues.redhat.com/browse/KEYCLOAK-13933 and add fix version 14? I hope we can address as a follow-up JIRA and in this PR to allow more work in parallel and unblock some other JIRAs, which depend on this PR

>     * Do we need the JSON view? IMHO it doesn't fit with the rest of the admin console. What is the use case for this? Do we expect the users to enter JSON directly? I'd say an import/export button would be a better alternative (like we do it for Clients e.g.). And if decide to input JSON through UI, we should do it properly and at least use some lib for JSON syntax highlighting and ideally with autocomplete as well.

Good question. I've did based on the design for the new admin console https://marvelapp.com/prototype/6e70eh2/screen/76353681 . However I am not 100% sure if we need it for the old admin console.

IMO we don't have time to do some great polishing and do some fancy improvements for make JSON editor very user-friendly (besides some low-hanging fruits like error handling as you pointed below). Could you please start thread on keycloak-dev about it or reply to existing thread related to client policies configuration https://groups.google.com/g/keycloak-dev/c/MfoTQbNPljE ?

> 
>     * If we keep the JSON view, we should probably improve the syntax error handling. Telling the user to open dev tools is not very user friendly. :) Maybe a modal dialog with more explicit error message would be better.

+1 :) Depends on answer on previous question. If we can do something easy to improve this, it will be great, but I am not yet 100% sure about this. Also if we do, we can possibly do it as a follow-up JIRA IMO?

> 
>     * Do we need built-in profiles? I suppose we do (FAPI and stuff), right? It looks a bit weird, the empty actions in the UI. :) Is it possible to enable/disable the built-in profiles? If yes, we could perhaps have a button for it in the actions column.

Yes, we need built-in profiles (renamed to "global" profiles in my latest changes on my laptop as discussed on keycloak-dev thread). Regarding the usability, is it possible again to create follow-up JIRA?

> 
>     * The executors for built-in profiles are read-only. This is of course correct. However, once you click it, there's no "back" button or something. It's a dead end. :)
> 
>     * Maybe we should have some bread-crumb navigation, like we have for clients and mappers e.g. It seems especially confusing for executors and conditions, when all the navigation (the Realm settings tabs) disappear. So maybe bread-crumb in this case could be somewhat helpful.

+1 . I've mentioned breadcrumbs in the possible TODO in the email I sent you yesterday though. But maybe we can do breadcrumbs and navigation improvements as a follow-up PR as well to allow have this PR ready earlier and then do follow-up work with additional improvements according to who has more time.

Another thing, which will be probably good to add are "Delete" confirmation (Currently stuff is deleted immediately after click "Delete" button, which is probably not ideal and it contradicsts the other places we have in Keycloak admin console). But maybe also doable as a follow-up task?

> 
>     * Just a nitpick. The "Realm Settings" button in the left nav bar is not highlighted as active.

Fixed already in my laptop. Feel free to create JIRA for this to not loose track of this.


> 
>     * It'd be a nice (and easy) improvement if client profiles could be opened directly from client policies. The linked profiles are currently not clickable.

Fixed already in my laptop. Feel free to create JIRA for this to not loose track of this.

> 
>     * I'm afraid we need some UI tests.

Yes, However adding tests take time and note there is new admin console in progress. So there won't be much time between client policies added in the old admin console and new admin console. Probably only one RHSSO release (RHSSO 7.5). Hence there is a question whether to spend time on UI tests considering this and considering the fact, that risk of regression is not very big and considering the fact that we have more places in the admin console, which are not tested?

However if you have time for help with admin console tests, it would be welcome :) But my vote is to start on them after previous points are addressed and UI is polished a bit to avoid some conflicts and need of rewrite etc...

> 
> 
> I know this is just a draft and you may (or may not :)) already have plans to fix/improve things I mentioned. And I'm also not sure if we even have time for stuff like that, or how good we want the UI to be. :) So please take it just as a recommendation/personal opinion from me. :)

Yes, exactly. We are quite constrained regarding time. I plan to add some improvements in the PR itself, but for the non-trivial things, we can hopefully fix them as a follow-up tasks? Also for JSON editor, I suggest to reply in the thread and possibly discuss on Monday tiger call whether to go with it or not...

vmuzikar commented:
@mposolda Created follow-up JIRAs based on the discussion here:
https://issues.redhat.com/browse/KEYCLOAK-18040
https://issues.redhat.com/browse/KEYCLOAK-18041
https://issues.redhat.com/browse/KEYCLOAK-18042
https://issues.redhat.com/browse/KEYCLOAK-18043
https://issues.redhat.com/browse/KEYCLOAK-18044

I also started a [thread](https://groups.google.com/g/keycloak-dev/c/iMu1oz6yFdM) on dev list about the JSON view.

vmuzikar reviewed (commented):
I went through the PR again with the latest changes. This time, I tried to look into the code as well (mainly in the backend, not that much in frontend). And I also tried to check the code as whole, not just the changes in this PR. I did not check line by line, though. :) It's almost impossible given the time frame and the size of the implementation.

The PR looks LGTM. Created one more follow-up JIRA ([KEYCLOAK-18079](https://issues.redhat.com/browse/KEYCLOAK-18079)) as the result of the [discussion](https://groups.google.com/g/keycloak-dev/c/iMu1oz6yFdM) on dev list (we're keeping the JSON view).

I noticed it's possible to create multiple conditions or executers of the same provider. Is it intentional. If so, we need to fix the form view as it displays only the first created.

One last thing we need to do is to run the UI tests to make sure there're no regressions.

A few generic question unrelated to this PR.
* I'm little bit concerned about performance. Do we have any results how it performs with hundreds or thousands of policies? I'm a little bit worried that we store everything in JSON, so there's even no room for optimization at the data layer (JPA, map store etc.). Also I noticed we don't use streams much in the implementation (e.g. for fetching roles etc.). This might hurt performance as well. But I'm not sure how critical this is for us. If the concern is even valid.
* Why do we use `@JsonIgnoreProperties(ignoreUnknown = true)` in some many places? I think it unnecessarily "loosens" the validation.
* I18n. I can see some hardcoded strings in conditions and executors but not sure if there's much we can do about it as it seems more of a general problem affecting e.g. client mappers too.

mposolda commented on core/src/main/java/org/keycloak/representations/idm/ClientProfilesRepresentation.java:
```diff
@@ -17,18 +17,25 @@
  * @author <a href="mailto:takashi.norimatsu.ws@hitachi.com">Takashi Norimatsu</a>
  */
-@JsonIgnoreProperties(ignoreUnknown = true)
+@JsonIgnoreProperties(ignoreUnknown = false)
```
Good point, will fix that.

mposolda commented on services/src/main/java/org/keycloak/services/resources/admin/ClientProfilesResource.java:
```diff
@@ -58,19 +61,24 @@ public ClientProfilesResource(RealmModel realm, AdminPermissionEvaluator auth) {
     @NoCache
     @Produces(MediaType.APPLICATION_JSON)
-    public String getProfiles() {
+    public ClientProfilesRepresentation getProfiles(@QueryParam("include-global-profiles") String includeGlobalProfiles) {
```
Good point, will fix that.

mposolda commented:
> I went through the PR again with the latest changes. This time, I tried to look into the code as well (mainly in the backend, not that much in frontend). And I also tried to check the code as whole, not just the changes in this PR. I did not check line by line, though. :) It's almost impossible given the time frame and the size of the implementation.
> 
> The PR looks LGTM. Created one more follow-up JIRA ([KEYCLOAK-18079](https://issues.redhat.com/browse/KEYCLOAK-18079)) as the result of the [discussion](https://groups.google.com/g/keycloak-dev/c/iMu1oz6yFdM) on dev list (we're keeping the JSON view).

Added fixVersion to Keycloak 14
> 
> I noticed it's possible to create multiple conditions or executers of the same provider. Is it intentional. If so, we need to fix the form view as it displays only the first created.

Ok, I will try to fix this in this PR

> 
> One last thing we need to do is to run the UI tests to make sure there're no regressions.
> 
> A few generic question unrelated to this PR.
> 
>     * I'm little bit concerned about performance. Do we have any results how it performs with hundreds or thousands of policies? I'm a little bit worried that we store everything in JSON, so there's even no room for optimization at the data layer (JPA, map store etc.). Also I noticed we don't use streams much in the implementation (e.g. for fetching roles etc.). This might hurt performance as well. But I'm not sure how critical this is for us. If the concern is even valid.

I don't we have time to do some serious performance testing and also I doubt that someone will have hundreds or thousands of policies/profiles in the single realm. But at the same time, totally agree that some performance improvement will be great and hopefully we can achieve it.

One thing, which should help, is usage of Amphibian, which I plan as a follow-up JIRA https://issues.redhat.com/browse/KEYCLOAK-17932 .

Regarding usage of streams, could you eventually create another follow-up JIRA for this, so we can possibly take a look at it as a follow-up?


> 
>     * Why do we use `@JsonIgnoreProperties(ignoreUnknown = true)` in some many places? I think it unnecessarily "loosens" the validation.

I think it is ok to remove those, so I will do it in this PR. I've kept them due the possible concerns during the migration (For example if type of some field in the "Configuration" will change), but I've checked that should not be a problem due the fact that during JSON read of ClientProfilesRepresentation (or ClientPoliciesRepresentation) the individual executors/conditions use ClientPolicyExecutorConfigurationRepresentation or ClientPolicyConditionConfigurationRepresentation. The specific configuration classes (for example SecureRequestObjectExecutor.Configuration) are used during the policy operations, but at that time, JSON should be migrated anyway to the new format.

FYI. There is also follow-up JIRA for handle migration from Keycloak 13 and make sure to add some migration tests https://issues.redhat.com/browse/KEYCLOAK-18069

> 
>     * I18n. I can see some hardcoded strings in conditions and executors but not sure if there's much we can do about it as it seems more of a general problem affecting e.g. client mappers too.

Yes, exactly. I won't care about those as other Keycloak places use it too and we might still need to do more refactoring of all places in the Keycloak in the future.

mposolda commented:
@vmuzikar Regarding the admin console tests, I've executed the pipeline https://keycloak-jenkins.com/job/universal-test-pipeline-server/377/ . There are same tests failing like with the master with both Chrome and FF, no regressions.

tnorimat commented on services/src/main/java/org/keycloak/services/clientpolicy/condition/ClientScopesConditionFactory.java:
```diff
@@ -40,10 +42,13 @@
         configProperties.add(property);
-        property = new ProviderConfigProperty(TYPE, "Scope Type", "Default or Optional", ProviderConfigProperty.LIST_TYPE, OPTIONAL);
+        property = new ProviderConfigProperty(TYPE, "Scope Type",
+                "If set to 'Default, condition will evaluate to true if client has some default scopes of the values specified by the 'Expected Scopes' property. " +
```
"If set to 'Default," might be "If set to 'Default',"

tnorimat commented on services/src/main/java/org/keycloak/services/clientpolicy/executor/SecureClientAuthEnforceExecutorFactory.java:
```diff
@@ -56,6 +54,25 @@ public void init(Scope config) {
     @Override
     public void postInit(KeycloakSessionFactory factory) {
+        ProviderConfigProperty isAugmentProperty = new ProviderConfigProperty(
+                IS_AUGMENT, "Augment Configuration", "If On, then the during client creation or update, the configuration of the client will be augmented to enforce usage of PKCE",
```
"enforce usage of PKCE" might be "enforce the authentication method to new clients" ?

mposolda committed:
KEYCLOAK-14209 Incorporate PR feedback

vmuzikar commented:
> Regarding usage of streams, could you eventually create another follow-up JIRA for this, so we can possibly take a look at it as a follow-up?

Done: https://issues.redhat.com/browse/KEYCLOAK-18089

tnorimat reviewed (commented):
@mposolda I've added some review comments. Could you check them?

mposolda commented:
@tnorimat I've incorporated all your comments. Thanks for the review!

@vmuzikar I've incorporated some of your feedback to this PR. The changes are:
- Form view is fixed to properly handle more executors/conditions of same type inside the profile/policy
- @JsonIgnoreProperties removed from all the places
- ClientProfilesResource (boolean variable)

For other things (performance, streams) we have follow-up JIRAs.

Do you have a chance to re-review?

vmuzikar reviewed (approved):
@mposolda Thanks for the update, LGTM (assuming the tests won't find any regression)!

mposolda commented:
@vmuzikar Thanks!
